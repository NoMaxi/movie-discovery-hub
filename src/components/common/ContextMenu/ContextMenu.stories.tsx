import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { ContextMenu } from "./ContextMenu";

interface ContextMenuAction {
    label: string;
    onClick: () => void;
}

const defaultMockActions: ContextMenuAction[] = [
    { label: "Edit", onClick: fn() },
    { label: "Delete", onClick: fn() },
];

const meta = {
    component: ContextMenu,
    argTypes: {
        isOpen: {
            control: "boolean",
            description: "Controls whether the context menu is visible",
            table: { defaultValue: { summary: 'false' } },
        },
        onClose: {
            description: "Callback function triggered when the menu requests to be closed (via close button or action click)",
        },
        actions: {
            control: "object",
            description: "Array of action objects ({ label: string, onClick: () => void })",
        },
        className: {
            control: "text",
            description: "Optional additional CSS classes for the container",
        },
        withCloseButton: {
            control: "boolean",
            description: "If true, shows a close (Ã—) button in the top-right corner",
            table: { defaultValue: { summary: 'false' } },
        },
    },
    args: {
        isOpen: true,
        actions: defaultMockActions,
        onClose: fn(),
    },
} satisfies Meta<typeof ContextMenu>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {},
};

export const InitiallyClosed: Story = {
    args: {
        isOpen: false,
    },
};

export const WithCloseButton: Story = {
    args: {
        withCloseButton: true,
    },
};

export const MultipleActions: Story = {
    args: {
        actions: [
            { label: "Copy Link", onClick: fn() },
            { label: "Share", onClick: fn() },
            { label: "Rename", onClick: fn() },
            { label: "Archive", onClick: fn() },
            { label: "Delete", onClick: fn() },
        ],
    },
};

export const WithCustomClasses: Story = {
    args: {
        className: "ring-2 ring-offset-2 ring-indigo-500 w-[200px]",
    },
};
