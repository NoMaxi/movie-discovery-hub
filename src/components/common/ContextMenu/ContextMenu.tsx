import { MouseEvent } from "react";

interface ContextMenuAction {
    label: string;
    onClick: () => void;
}

interface ContextMenuProps {
    isOpen: boolean;
    onClose: () => void;
    actions: ContextMenuAction[];
    className?: string;
    withCloseButton?: boolean;
}

export const ContextMenu = ({
    isOpen,
    onClose,
    actions,
    className = "",
    withCloseButton = false,
}: ContextMenuProps) => {
    if (!isOpen) {
        return null;
    }

    const handleActionClick = (actionOnClick: () => void) => {
        actionOnClick();
        onClose();
    };

    const handleMenuClick = (e: MouseEvent) => {
        e.stopPropagation();
    };

    const handleMenuClose = (e: MouseEvent) => {
        e.stopPropagation();
        onClose();
    };

    return (
        <div
            data-testid="context-menu"
            className={`
                context-menu
                absolute z-20 min-w-[100px]
                flex flex-col
                bg-[var(--color-content-background)]
                ring-1 ring-black shadow-lg rounded
                ${className}
            `}
            onClick={handleMenuClick}
        >
            {withCloseButton && (
                <button
                    data-testid="context-menu-close-button"
                    className="
                    context-menu-close-button
                    absolute top-1 right-2
                    text-xl leading-none text-[var(--color-gray-lighter)]
                    hover:text-[var(--color-text)]
                    cursor-pointer
                "
                    onClick={handleMenuClose}
                >
                    &times;
                </button>
            )}

            <ul className="m-0 p-0 list-none">
                {actions.map((action) => (
                    <li key={action.label}>
                        <button
                            data-testid={`context-menu-${action.label.toLowerCase()}-button`}
                            className={`
                                context-menu-${action.label.toLowerCase()}-button
                                block w-full text-left px-4 py-2
                                text-base text-[var(--color-text)]
                                hover:bg-[var(--color-primary)]
                                transition-colors duration-150
                                cursor-pointer
                            `}
                            onClick={() => handleActionClick(action.onClick)}
                        >
                            {action.label}
                        </button>
                    </li>
                ))}
            </ul>
        </div>
    );
};
