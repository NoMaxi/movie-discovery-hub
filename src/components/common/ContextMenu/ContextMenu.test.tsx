import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { ContextMenu } from "./ContextMenu";

interface ContextMenuAction {
    label: string;
    onClick: () => void;
}

describe("ContextMenu", () => {
    let onCloseMock: jest.Mock;
    let user: UserEvent;
    let mockActions: ContextMenuAction[];

    beforeEach(() => {
        onCloseMock = jest.fn();
        mockActions = [
            { label: "Edit", onClick: jest.fn() },
            { label: "Delete", onClick: jest.fn() },
        ];
        user = userEvent.setup();
    });

    const renderComponent = (isOpen: boolean, props: Partial<React.ComponentProps<typeof ContextMenu>> = {}) => {
        const defaultProps: React.ComponentProps<typeof ContextMenu> = {
            isOpen,
            onClose: onCloseMock,
            actions: mockActions,
        };
        const mergedProps = { ...defaultProps, ...props };
        return render(<ContextMenu {...mergedProps} />);
    };

    test("Should not render when 'isOpen' prop is false", () => {
        renderComponent(false);
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
    });

    test("Should render correctly without close button by default when 'isOpen' is true", () => {
        const { asFragment } = renderComponent(true, { className: "custom-class" });
        const menuElement = screen.getByTestId("context-menu");

        expect(menuElement).toBeInTheDocument();
        expect(menuElement).toHaveClass("custom-class");
        expect(screen.queryByTestId("context-menu-close-button")).not.toBeInTheDocument();
        expect(screen.getByTestId("context-menu-edit-button")).toBeInTheDocument();
        expect(screen.getByText("Edit")).toBeInTheDocument();
        expect(screen.getByTestId("context-menu-delete-button")).toBeInTheDocument();
        expect(screen.getByText("Delete")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot("without-close-button");
    });

    test("Should render correctly with close button when 'withCloseButton' is true", () => {
        const { asFragment } = renderComponent(true, { withCloseButton: true });
        const menuElement = screen.getByTestId("context-menu");

        expect(menuElement).toBeInTheDocument();
        expect(screen.getByTestId("context-menu-close-button")).toBeInTheDocument();
        expect(screen.getByTestId("context-menu-edit-button")).toBeInTheDocument();
        expect(screen.getByTestId("context-menu-delete-button")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot("with-close-button");
    });

    test("Should call 'onClose' prop when close button is clicked", async () => {
        renderComponent(true, { withCloseButton: true });
        const closeButton = screen.getByTestId("context-menu-close-button");

        await user.click(closeButton);

        expect(onCloseMock).toHaveBeenCalledTimes(1);
        expect(mockActions[0].onClick).not.toHaveBeenCalled();
        expect(mockActions[1].onClick).not.toHaveBeenCalled();
    });

    test("Should call correct action's 'onClick' and 'onClose' props when action button is clicked", async () => {
        renderComponent(true);
        const editButton = screen.getByTestId("context-menu-edit-button");

        await user.click(editButton);

        expect(mockActions[0].onClick).toHaveBeenCalledTimes(1);
        expect(mockActions[1].onClick).not.toHaveBeenCalled();
        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    test("Should stop propagation when clicking inside the menu container", async () => {
        const handleOuterClick = jest.fn();
        render(
            <div data-testid="outer-container" onClick={handleOuterClick}>
                <ContextMenu isOpen={true} onClose={onCloseMock} actions={mockActions} />
            </div>,
        );

        const menuElement = screen.getByTestId("context-menu");
        await user.click(menuElement);

        expect(handleOuterClick).not.toHaveBeenCalled();
        expect(onCloseMock).not.toHaveBeenCalled();
    });

    test("Should stop propagation when clicking the close button", async () => {
        const handleOuterClick = jest.fn();
        render(
            <div data-testid="outer-container" onClick={handleOuterClick}>
                <ContextMenu isOpen={true} onClose={onCloseMock} actions={mockActions} withCloseButton={true} />
            </div>,
        );

        const closeButton = screen.getByTestId("context-menu-close-button");
        await user.click(closeButton);

        expect(handleOuterClick).not.toHaveBeenCalled();
        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    test("Should stop propagation when clicking an action button", async () => {
        const handleOuterClick = jest.fn();
        render(
            <div data-testid="outer-container" onClick={handleOuterClick}>
                <ContextMenu isOpen={true} onClose={onCloseMock} actions={mockActions} />
            </div>,
        );

        const editButton = screen.getByTestId("context-menu-edit-button");
        await user.click(editButton);

        expect(handleOuterClick).not.toHaveBeenCalled();
        expect(mockActions[0].onClick).toHaveBeenCalledTimes(1);
        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });
});
