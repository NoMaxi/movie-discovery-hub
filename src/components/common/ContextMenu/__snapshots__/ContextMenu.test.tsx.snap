// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu Should render correctly with close button when 'withCloseButton' is true: with-close-button 1`] = `
<DocumentFragment>
  <div
    class="
                context-menu
                absolute z-20 min-w-[100px]
                flex flex-col justify-center
                bg-[var(--color-background)]
                ring-1 ring-black shadow-lg rounded
                overflow-hidden
                
            "
    data-testid="context-menu"
  >
    <button
      class="
                    context-menu-close-button
                    absolute top-1 right-2
                    text-xl leading-none text-[var(--color-gray-lighter)]
                    hover:text-[var(--color-text)]
                    cursor-pointer
                "
      data-testid="context-menu-close-button"
    >
      Ã—
    </button>
    <ul>
      <li>
        <button
          class="
                                context-menu-edit-button
                                block w-full text-left px-4 py-2
                                text-base text-[var(--color-text)]
                                hover:bg-[var(--color-primary)]
                                transition-colors duration-150
                                cursor-pointer
                            "
          data-testid="context-menu-edit-button"
        >
          Edit
        </button>
      </li>
      <li>
        <button
          class="
                                context-menu-delete-button
                                block w-full text-left px-4 py-2
                                text-base text-[var(--color-text)]
                                hover:bg-[var(--color-primary)]
                                transition-colors duration-150
                                cursor-pointer
                            "
          data-testid="context-menu-delete-button"
        >
          Delete
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ContextMenu Should render correctly without close button by default when 'isOpen' is true: without-close-button 1`] = `
<DocumentFragment>
  <div
    class="
                context-menu
                absolute z-20 min-w-[100px]
                flex flex-col justify-center
                bg-[var(--color-background)]
                ring-1 ring-black shadow-lg rounded
                overflow-hidden
                custom-class
            "
    data-testid="context-menu"
  >
    <ul>
      <li>
        <button
          class="
                                context-menu-edit-button
                                block w-full text-left px-4 py-2
                                text-base text-[var(--color-text)]
                                hover:bg-[var(--color-primary)]
                                transition-colors duration-150
                                cursor-pointer
                            "
          data-testid="context-menu-edit-button"
        >
          Edit
        </button>
      </li>
      <li>
        <button
          class="
                                context-menu-delete-button
                                block w-full text-left px-4 py-2
                                text-base text-[var(--color-text)]
                                hover:bg-[var(--color-primary)]
                                transition-colors duration-150
                                cursor-pointer
                            "
          data-testid="context-menu-delete-button"
        >
          Delete
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
